src/devices/vdml_link.c

```diff
@@ -0,0 +1,234 @@
+    }
+    uint32_t rtv = vexDeviceGenericRadioReceive(device->device_info, (uint8_t*)dest, data_size);
+    return_port(port - 1, rtv);
+}
+
+uint32_t link_transmit(uint8_t port, void* data, uint16_t data_size) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    if(!vexDeviceGenericRadioLinkStatus(device->device_info)) {
+        errno = ENXIO;
+        return_port(port - 1, PROS_ERR);
+    }
+    if(data_size + PROTOCOL_SIZE > vexDeviceGenericRadioWriteFree(device->device_info)) {
+        errno = EBUSY;
+        kprintf("Data Size too big: %d\n", data_size + PROTOCOL_SIZE);
+        return_port(port - 1, 0);
+    }
+    if(data == NULL) {
+        errno = EINVAL;
+        return_port(port - 1, PROS_ERR);
+    }
+    // calculated checksum
+    uint8_t checksum = start_byte;
+    uint8_t size_tx_buf[2];
+    size_tx_buf[1] = (data_size >> 8) & 0xff;
+    size_tx_buf[0] = (data_size) & 0xff;
+    checksum ^= size_tx_buf[1];
+    checksum ^= size_tx_buf[0];
+    
+    for(int i = 0; i < data_size; i++) {
+        checksum ^= ((uint8_t*)data)[i];
+    }
+
+    // send protocol
+    vexDeviceGenericRadioTransmit(device->device_info, &start_byte, 1);
+    vexDeviceGenericRadioTransmit(device->device_info, size_tx_buf, 2);
+    vexDeviceGenericRadioTransmit(device->device_info, (uint8_t*)data, data_size);
+    vexDeviceGenericRadioTransmit(device->device_info, &checksum, 1);
+    return_port(port - 1, 1);
+}
+
+uint32_t link_receive(uint8_t port, void* data, uint16_t data_size) {
+    claim_port_i(port - 1, E_DEVICE_SERIAL);
+    if(!vexDeviceGenericRadioLinkStatus(device->device_info)) {
+        errno = ENXIO;
+        return_port(port - 1, PROS_ERR);
+    }
+    if(data == NULL) {
+        errno = EINVAL;
+        return_port(port - 1, PROS_ERR);
+    }
+    if(data_size + PROTOCOL_SIZE > vexDeviceGenericRadioReceiveAvail(device->device_info)) {
+        _clear_rx_buf(device);
+        errno = EBUSY;
+        return_port(port - 1, 0);
+    }
+    uint8_t header_byte;
+    uint8_t received_size;
+    // process protocol
+    received_size = vexDeviceGenericRadioReceive(device->device_info, &header_byte, 1); // 0x33
+    if(start_byte != header_byte || received_size != 1) {
+        kprintf("[VEXLINK] Invalid Header Byte Received Port %d, header byte: %x", port, header_byte);
+        errno = EBADMSG;
+        return_port(port - 1, PROS_ERR);
+    }
+    // datasize
+    uint16_t _size;
+    received_size = vexDeviceGenericRadioReceive(device->device_info, (uint8_t*)&_size, 2);
+    if(received_size != 2 || _size != data_size) {
+        _clear_rx_buf(device);
+        kprintf("[VEXLINK] Invalid Data Size (Size: %d ) Received Port %d, flushing RX buffer!\n", _size, port);
+        errno = EBADMSG;
+        return_port(port - 1, PROS_ERR);
+    }
+    // receive data
+    uint32_t rtv;
+    rtv = vexDeviceGenericRadioReceive(device->device_info, (uint8_t*)data, data_size);
+    if(rtv != data_size || _size != data_size) {
```

Comment:
It's an error because if we didn't get the right amount of bytes from the retrieval, we could've accidentally also retrieved the checksum or any future packets, meaning that we'll have a malformed packet that we can't really fix since VEX only provided an opaque buffer. In that case, we clear the buffer so any new packets will be easier to retrieve.